### DISCUSSION TITLE
Enhancing Helm Chart Security and Compatibility with OpenShift

### DISCUSSION THREAD

**Discussion by elysweyr** [nextcloud/all-in-one#4364] [Ideas] [answered]
**Background:**
In the realm of Kubernetes deployments, Helm charts play a pivotal role in simplifying application packaging and deployment. However, there's an imperative need to reassess security practices within Helm charts, specifically in the context of init-containers and container user permissions. Currently, the prevalent use of `chmod 777` in init-containers and static User IDs (UIDs) within containers raises security concerns and is inconsistent with the security constraints enforced by platforms like OpenShift.

**Objective:**
This proposal aims to enhance the security posture of Helm charts by eliminating the use of `chmod 777` in init-containers and evaluating the feasibility of transitioning from static UIDs to dynamic, randomly assigned UIDs. By aligning with OpenShift's security context constraints default behavior, this enhancement ensures a more robust and secure deployment environment.

**Key Components:**

**Removal of chmod 777 in Init-Containers:**

Update Helm chart templates to eliminate the usage of chmod 777 in init-containers.
Adopt more granular and controlled file permissions based on the principle of least privilege.
**Transition from Static to Random UIDs:**

Evaluate the existing containers with static UIDs and assess the feasibility of introducing random UID assignment.
Implement a mechanism to dynamically assign UIDs during container runtime, enhancing security by reducing predictability.
**OpenShift Security Context Constraints Compatibility:**

Align the Helm chart with OpenShift's security context constraints default behavior.
Ensure that security contexts within the Helm chart are compatible with OpenShift's security policies and do not compromise the platform's security model.
**Documentation and Best Practices:**

Update Helm chart documentation to reflect the revised security practices.
Provide guidelines and best practices for secure init-container configurations and UID assignments.
Clearly communicate the reasons behind these changes and the potential impact on existing deployments.

**Benefits:**

Improved security posture by eliminating insecure file permissions (chmod 777) in init-containers.
Enhanced security through the adoption of dynamic UID assignment, reducing predictability.
Alignment with OpenShift's security context constraints default behavior, ensuring compatibility and adherence to platform security policies.
Rationale:
OpenShift, by default, enforces security constraints that restrict the use of overly permissive file permissions and advocate for dynamic UID assignment. This proposal seeks to align Helm charts with such security standards, addressing potential vulnerabilities and promoting secure deployment practices.

By implementing these changes, Helm charts will not only enhance security but also foster compatibility with OpenShift, providing users with a more robust and consistent experience across Kubernetes platforms. This feature proposal aligns with industry best practices and security standards, contributing to a safer and more resilient containerized application ecosystem.

Docu: https://docs.openshift.com/container-platform/4.15/openshift_images/create-images.html#use-uid_create-images

**Comment by szaimen** [comment]
Hi, I fear this is out-of-scope for AIOs Helm chart for now.

**Comment by royalhorror** [comment]
Hi
Is this still out of scope? OpenShift is one of the most widely used Kubernetes platforms, and the security constraints in OpenShift are best-practice. I find it strange that Nextcloud, a platform that is often used to store sensitive data, doesn't adhere to at least some of these practices regardless of OpenShift.

**Reply by szaimen** [reply, marked as answer]
Hi, for an enterprise-ready and scalable deployment method based on Helm Charts (also available for Podman and OpenShift), please [contact Nextcloud GmbH](https://nextcloud.com/enterprise/).

**Reply by royalhorror** [reply]
Hi @szaimen 
Do you view Podman and OpenShift/OKD as enterprise use only platforms?
I wouldn't even say Podman is a platform, as it doesn't have clustering functionality. IMO Podman is just a good alternative to Docker and neither Podman nor Docker are enterprise platforms.

OpenShift is definitively an enterprise platform in many configurations, but you also have Single Node OpenShift and MicroShift, which in the context of hosting Nextcloud is not an enterprise solution alone. Both Single Node OpenShift and MicroShift (initially created for edge deployments) are great for home deployments and home labs.

With the license for Nextcloud Enterprise starting at a minimum of 100 users, I can't justify or afford that for a home deployment for two people.
